fix(tests): fix malloc assertion error by replacing thread tcp server.

zhongweiy opened the pull request:
Due to many OpenSSL objects are not thread-safe (https://www.openssl.org/docs/faq.html) and it causes intermittent crash when those objects are used in thread based tcp server. Replace it with process based tcp server resolves this error.

This change only replaces the thread tcp server usage in 11-reports_spec.lua file to keep the scope small.

Before this fix, crash like "malloc_consolidate(): unaligned fastbin chunk detected." happens when running spec/01-unit/11-reports_spec.lua about 200 times.

After this fix, there is no crash after running the same test suites more than 5000 times.

<!--
NOTE: Please read the CONTRIBUTING.md guidelines before submitting your patch,
and ensure you followed them all:
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributing

Refer to the Kong Gateway Community Pledge to understand how we work
with the open source community:
https://github.com/Kong/kong/blob/master/COMMUNITY_PLEDGE.md
-->

### Summary

<!--- Why is this change required? What problem does it solve? -->

### Checklist

- [ ] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

<!--- If it fixes an open issue, please link to the issue here. -->
Fix #_[KAG-3241]_


[KAG-3241]: https://konghq.atlassian.net/browse/KAG-3241?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

bungle commented:
@zhongweiy is there reason we just don't use the mock server?
https://github.com/Kong/kong/blob/master/spec/fixtures/template_inject/nginx_kong_test_custom_inject_stream.lua#L1-L32

zhongweiy commented:
@bungle I have updated with the mock server in kong with customized config. The code looks more clear. But the single test case running time increases a lot:

thread tcp_server : 30ms
standalone nginx : 130ms
kong mock server : 2000ms

There are about 50 usages of tcp server in our tests code. So after we replace all thread tcp_server with kong mock server, there will be about 2*50 = 100 seconds slow down. Consider we have split the tests in CI, the real slowdown is probably less than 100 seconds.

What do you think?

bungle commented:
> thread tcp_server : 30ms (the current mainline)
> standalone nginx : 130ms (the first version of this PR)
> kong mock server : 2000ms (the latest version of this PR)

What are the numbers? Is this a startup time? Most of the time you don't need to start `kong mock server`  as it is already started as part of starting Kong. So you don't need it separate as you are already testing Kong. It is just another server block in that same process.

zhongweiy commented:
> > thread tcp_server : 30ms (the current mainline)
> > standalone nginx : 130ms (the first version of this PR)
> > kong mock server : 2000ms (the latest version of this PR)
> 
> What are the numbers? Is this a startup time? 

The number is the whole time of running a test. i.e. 30ms = thread tcp_server start time + test run time. The test run time should be small. So in 130ms, most of the time (100+ ms) is from standalone nginx start time. In 2000ms, 1970+ ms is from kong start time.

>Most of the time you don't need to start `kong mock server` as it is already started as part of starting Kong (that you are usually about to test). So you don't need it separate as you are already testing Kong. It is just another server block in that same process.

Do you mean the case like in [some lazy_setup/setup](https://github.com/Kong/kong/blob/1a865512d90725e5675ae773e2c5c4d2ebf362bd/spec/03-plugins/27-aws-lambda/99-access_spec.lua#L505) that start_kong, and the test case in the describe block shares the kong? If yes, then we can move the `start_kong` to the `setup` reduce the overall run time. However, this is not exactly the same as the origin test, which start thread tcp server for each test case instead of setting up it in the `setup` block.

bungle commented:
> Do you mean the case like in [some lazy_setup/setup](https://github.com/Kong/kong/blob/1a865512d90725e5675ae773e2c5c4d2ebf362bd/spec/03-plugins/27-aws-lambda/99-access_spec.lua#L505) that start_kong, and the test case in the describe block shares the kong? If yes, then we can move the `start_kong` to the `setup` reduce the overall run time. However, this is not exactly the same as the origin test, which start thread tcp server for each test case instead of setting up it in the `setup` block.

@zhongweiy,

I think that start server for each test was probably because of limitation of that server. I think it should be possibly to write it so that you don't need to restart it all the time. What do you think?

zhongweiy commented:
> > Do you mean the case like in [some lazy_setup/setup](https://github.com/Kong/kong/blob/1a865512d90725e5675ae773e2c5c4d2ebf362bd/spec/03-plugins/27-aws-lambda/99-access_spec.lua#L505) that start_kong, and the test case in the describe block shares the kong? If yes, then we can move the `start_kong` to the `setup` reduce the overall run time. However, this is not exactly the same as the origin test, which start thread tcp server for each test case instead of setting up it in the `setup` block.
> 
> @zhongweiy,
> 
> I think that start server for each test was probably because of limitation of that original tcp server (or copy paste programming). I think it should be possible to write it so that you don't need to restart it all the time. At least I don't see any reason to restart server all the time. What do you think?

Got it. I agree starting only one server for each test file will not impact the correctness of the test. The small downside is the `get_echo_server_received_data` function will be a little more complex. But it's not a big issue. Let me update it.

chobits commented on changelog/unreleased/kong/fix_malloc_assertion_error.yml:
```diff
@@ -0,0 +1,3 @@
+message: "fix malloc assertion error by replacing thread tcp server in reports_spec.lua."
+type: bugfix
+scope: Core
```
@ADD-SP Is any modifcation in `spec` about test cases uncessary to write one changelog entry?

I noticed that in my previous PR, you comment that we dont need this in test cases.

ADD-SP replied:
Only the user-facing changes need the changlog.

ADD-SP commented on changelog/unreleased/kong/fix_malloc_assertion_error.yml:
These changes are not user-facing, we don't need the changelog.

zhongweiy commented:
@bungle update it to share the tcp server between tests.

the 11-reports_spec.lua test file runtime change:

1. server shared version: 2.6 seconds
2. server not shared version: 40 seconds
3. the mainline thread tcp server version: 0.5 seconds

bungle commented on spec/fixtures/template_inject/nginx_kong_test_tcp_echo_server_custom_inject_stream.lua:
```diff
@@ -0,0 +1,23 @@
+> end
+    ssl_protocols TLSv1.1 TLSv1.2 TLSv1.3;
+
+    content_by_lua_block {
+        local sock = assert(ngx.req.socket())
+        local data = sock:receive()  -- read a line from downstream
+        if data then
+            sock:send(data.."\n") -- echo whatever was sent
+            ngx.log(ngx.INFO, "received data: " .. data)
+        else
+            ngx.log(ngx.WARN, "Nothing received")
+        end
+    }
```
Is there a reason why we cannot just use this:
https://github.com/Kong/kong/blob/master/spec/fixtures/template_inject/nginx_kong_test_custom_inject_stream.lua#L13-L29

Or modify that to support also the reports tests?

It looks like both are kinda echo servers with minor differences.

zhongweiy replied:
I was thinking for making the scope small, this PR tries to not touch existing functionality if possible. 

And another reason is there are another 47 tcp_server() usage needs to be removed in our testing. And there may be other required modification on this `nginx_kong_test_tcp_echo_server_custom_inject_stream.lua` file. To enable the later on change be separated from existing one, this PR puts the change in the new file.

bungle replied:
As discussed offline, this is good, and I agree with your points here.

bungle reviewed (changes requested):
Can the current tcp/udp mock server work here?

bungle commented on spec/01-unit/11-reports_spec.lua:
```diff
@@ -2,11 +2,32 @@ local meta = require "kong.meta"
   local port = 8189
-  local opts = { tls = true }
+
+  setup(function()
```
```suggestion
  lazy_setup(function()
```

zhongweiy replied:
will update.

bungle commented on spec/01-unit/11-reports_spec.lua:
```diff
@@ -2,11 +2,32 @@ local meta = require "kong.meta"
+    assert(helpers.is_echo_server_ready())
+  end)
+
+  teardown(function()
```
```suggestion
  lazy_teardown(function()
```

zhongweiy replied:
will update

bungle commented on spec/01-unit/11-reports_spec.lua:
```diff
@@ -2,11 +2,32 @@ local meta = require "kong.meta"
+               log_level = "info",
+                      }))
+
+    assert(helpers.is_echo_server_ready())
```
I have not seen us using `is_kong_<any>_server_ready` in a past. Is there reason for this? Or how it works elsewhere in tests where we just start kong? Is it covered by timeout of admin/proxy client in helpers? But shouldn't it then be covered with reports client too?

zhongweiy replied:
In the original thread based `tcp_server()`, there is a block of code that does the same function to check whether the echo server ready.

bungle reviewed (commented):
Question
